openedx_tagging/core/tagging/import_export/actions.py

```diff
@@ -0,0 +1,444 @@
+
+    def __init__(self, taxonomy: Taxonomy, tag, index: int):
+        self.taxonomy = taxonomy
+        self.tag = tag
+        self.index = index
+
+    def __repr__(self):
+        return str(_(f"Action {self.name} (index={self.index},id={self.tag.id})"))
+
+    def __str__(self):
+        return self.__repr__()
+
+    @classmethod
+    def valid_for(cls, taxonomy: Taxonomy, tag) -> bool:
+        """
+        Implement this to meet the conditions that a `TagItem` needs
+        to have for this action. All actions that are valid with
+        this function are created.
+        """
+        raise NotImplementedError
+
+    def validate(self, indexed_actions) -> List[ImportActionError]:
+        """
+        Implement this to find inconsistencies with tags in the
+        database or with previous actions.
+        """
+        raise NotImplementedError
+
+    def execute(self):
+        """
+        Implement this to execute the action.
+        """
+        raise NotImplementedError
+
+    def _get_tag(self):
+        """
+        Returns the respective tag of this actions
+        """
+        return self.taxonomy.tag_set.get(external_id=self.tag.id)
+
+    def _search_action(
+        self,
+        indexed_actions: dict,
+        action_name: str,
+        attr: str,
+        search_value: str,
+    ):
+        """
+        Use this function to find and action using an `attr` of `TagItem`
+        """
+        for action in indexed_actions[action_name]:
+            if search_value == getattr(action.tag, attr):
+                return action
+
+        return None
+
+    def _validate_parent(self, indexed_actions) -> ImportActionError:
+        """
+        Validates if the parent is created
+        """
+        try:
+            # Validates that the parent exists on the taxonomy
+            self.taxonomy.tag_set.get(external_id=self.tag.parent_id)
+        except Tag.DoesNotExist:
+            # Or if the parent is created on previous actions
+            if not self._search_action(
+                indexed_actions, CreateTag.name, "id", self.tag.parent_id
+            ):
+                return ImportActionError(
+                    action=self,
+                    tag_id=self.tag.id,
+                    message=_(
+                        f"Unknown parent tag ({self.tag.parent_id}). "
+                        "You need to add parent before the child in your file."
+                    ),
+                )
+
+    def _validate_value(self, indexed_actions):
+        """
+        Check for value duplicates in the models and in previous create/rename
+        actions
+        """
+        try:
+            # Validates if exists a tag with the same value on the Taxonomy
+            tag = self.taxonomy.tag_set.get(value=self.tag.value)
+            return ImportActionError(
+                action=self,
+                tag_id=self.tag.id,
+                message=_(f"Duplicated tag value with tag (id={tag.id})."),
```

Comment:
Yes, you have made me realize that `id` is not useful to the user, but `external_id` is, since it is something he handles. I've updated some posts that used `id` to use `external_id` 6a472e3b6c7f458d1aeebbc7814b92bf4e0d66c8

Reply:
Yes, thanks! That's a nice improvement. I can see already from the updated test cases that the error messages are more useful. 

![Screenshot 2023-08-10 at 4 09 52 PM](https://github.com/openedx/openedx-learning/assets/945577/db066b7b-a230-4414-b31a-7358b55aa8f6)